cybergis_compute_client/CyberGISCompute.py

```diff
@@ -66,7 +68,87 @@ def __init__(self, url="cgjobsup.cigi.illinois.edu", port=443, protocol='HTTPS',
         self.job = None
         self.recentDownloadPath = None
 
-    def login(self, manualLogin=True, verbose=True):
+    def encrypt_token(self, token):
+        """
+        Encrypts the token using host variable.
+
+        Args:
+            token (str): User/Environment provided token.
+        """
+        self.jupyterhubApiToken = base64.b64encode((self.jupyterhubHost + '@' + token).encode('ascii')).decode('utf-8')
+
+    def get_jupyterhubHost(self):
+        """
+        Gets the jupyterhub host(str) from the user.
+        """
+        if (self.jupyterhubHost is None):
+            print("Please copy the JupyterHub url along with port. E.g http://127.0.0.1:8081")
+            self.jupyterhubHost = input('Enter your jupyterhubHost here: ')
+
+    def set_username(self):
+        """
+        Authenticates the token(str) and saves the username(str).
+        """
+        res = self.client.request('GET', '/user', {"jupyterhubApiToken": self.jupyterhubApiToken})
+        self.username = res['username']
+
+    def save_token(self):
+        """
+        Writes token(str) to json file.
+        """
+        with open('./cybergis_compute_user.json', 'w') as json_file:
+            json.dump({"token": self.jupyterhubApiToken}, json_file)
+
+    def login_token(self):
+        """
+        Saves username(str) and token(str).
+        """
+        try:
+            self.set_username()
+            self.save_token()
+            return self.login()
+        except:
+            print('❌ Failed to login via system token')
+
+    def login_manual(self):
+        """
+        Asks for token and host from user and calls login_token function.
+        """
+        if self.isJupyter:
+            print('📢 Please go to Control Panel -> Token, request a new API token')
+            token = getpass.getpass('Enter your API token here')
+            try:
+                self.get_jupyterhubHost()
```

Comment:
These three lines:

```
self.get_jupyterhubHost()
self.encrypt_token(envToken)
return self.login_token()
```

get repeated in a few places in the same order, so we may want to just turn it into a function of it's own for readability and maintainability. Not urgent, but something we should back to before last release.

Reply:
Done !